src/fairseq2/recipes/hg/asr_eval.py

```diff
@@ -122,77 +150,95 @@ def _librispeech_asr_to_batch(examples: Example) -> Seq2SeqBatch:
     )
 
 
-@lru_cache(maxsize=None)
-def get_cached_tokenizer(tokenizer_name: str) -> TextTokenizer:
-    return load_text_tokenizer(tokenizer_name)
-
-
-def _preprocess_example(
-    example: Example, tokenizer_name: str, device: torch.device
-) -> Example:
+def prepare_dataset(
+    config: AsrEvalConfig, processor: Optional[Callable[[Example], Example]] = None
+) -> Dataset:
     """
-    Preprocesses an individual example by converting the audio array to a PyTorch tensor
-    and encoding the text.
+    Prepares a dataset for evaluation. The dataset is loaded from the
+    HF datasets and preprocessed using the provided processor.
 
     Args:
-        example (dict): A dictionary containing "audio" and "text" keys.
-        tokenizer_name (str): The name of the tokenizer to use.
-        device (torch.device): The device to store the tensors.
+        config (AsrEvalConfig): The configuration for the evaluation.
+        processor (Callable): A function to preprocess examples.
 
     Returns:
-        dict: A dictionary with "audio" and "text" as PyTorch tensors.
+        Dataset: The prepared dataset.
     """
-    tokenizer = get_cached_tokenizer(tokenizer_name)
-    encoder = tokenizer.create_encoder(device=device)
-    audio_tensor = (
-        torch.from_numpy(example["audio"]["array"]).to(torch.float16).to(device)
+    iterable_ds = load_dataset(config.dataset_name, split=config.split, streaming=True)
+    ds = Dataset.from_generator(
+        lambda: itertools.islice(iterable_ds, 0, config.max_samples),
+        features=iterable_ds.features,
     )
-    text_tensor = encoder(example["text"].lower()).to(device)
-    return {"audio": audio_tensor, "text": text_tensor}
+    ds = ds.map(lambda x: extract_features(x))
+
+    if processor is not None:
+        ds = ds.map(processor)
+
+    format = {
+        "type": "torch",
+        "format_kwargs": {"dtype": config.dtype},
+    }
+    ds.set_format(**format, columns=["audio", "text"])
+
+    return ds
 
 
-def seq2seq_preprocessor(batch: Seq2SeqBatch) -> tuple[SequenceBatch, SequenceBatch]:
+def evaluator_preprocessor(batch: Seq2SeqBatch) -> tuple[SequenceBatch, SequenceBatch]:
     return SequenceBatch(batch.source_seqs, batch.source_padding_mask), SequenceBatch(
         batch.target_seqs, batch.target_padding_mask
     )
 
 
-def postprocesser(
+def evaluator_postprocesser(
     outputs: Any, targets: SequenceBatch, tokenizer: TextTokenizer
 ) -> tuple[list[str], list[str]]:
     decoder = tokenizer.create_decoder()
     pad_idx = tokenizer.vocab_info.pad_idx
 
     hypotheses, _ = outputs.generate_hypotheses(pad_idx=pad_idx)
     predictions = [decoder(item) for item in hypotheses]
-    references = [decoder(item) for item in targets.seqs.to(torch.int32)]
+    references = cast(list[str], targets.seqs)
```

Comment:
Are you sure all targets.seqs will have torch.int32 ?

Reply:
At this point target seqs are just the strings themselves.

I have removed the encoding and decoding of text since its redundant (text is being encoded only for it to be decoded again without using the tokens)